Splitting string into Alpha-Num tokens with SIMD intrinsics.

Enmk opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Using SSE2 or SSE4.2 SIMD intrinsics to speed up tokenization in bloom filters.
...

Enmk committed:
Initial implementation of splitting string into Alpha-Num tokens with SIMD intrinsics.

alexey-milovidov commented on dbms/src/Storages/MergeTree/MergeTreeIndexFullText.cpp:
```diff
@@ -606,8 +606,60 @@ bool SplitTokenExtractor::next(const char * data, size_t len, size_t * pos, size
+
     while (*pos < len)
     {
+#if __SSE2__
```
#if defined(__SSE2__)

Enmk replied:
Yep, fixed that and some other run-time issues too.

Enmk committed:
SplitTokenExtractor::next unit-tests

Enmk committed:
Fixed compilation issues and fixed several bugs in SplitTokenExtractor::next

* Handling all characters above 0x80 as symbols (fixes UTF8 tokens)
* Properly handling tokens that end exactly on haystack boundary.

Enmk committed:
Minor: excluded superflous includes

Enmk committed:
Fixed final garbage token

Also, added getTrailingZeroBitsUnsafe(), updated comments and cleaned up
test cases

Enmk committed:
Merge commit 'cd88b5380c886e6d4f84d0b29d9a27e1a12d2cbd' into SIMD_tokenization

danlark1 commented on dbms/src/Storages/tests/gtest_SplitTokenExtractor.cpp:
```diff
@@ -0,0 +1,174 @@
+
+struct SplitTokenExtractorTestCase
+{
+    const char * description;
```
nit: please, use std::string_view as operator<< is not going to call strlen each time

Enmk replied:
done

danlark1 commented on dbms/src/Storages/tests/gtest_SplitTokenExtractor.cpp:
```diff
@@ -0,0 +1,174 @@
+
+    for (const auto & expected_token : param.tokens)
+    {
+        SCOPED_TRACE(++i);
```
Why not SCOPED_TRACE(description + " " + token), I don't see how numbers make the life easier in detecting the test case

Enmk replied:
since description and token are shown anyway in failed tests report, in case of similar tokens it pays to know what was the token index.

danlark1 replied:
Ok

danlark1 commented on dbms/src/Storages/tests/gtest_SplitTokenExtractor.cpp:
```diff
@@ -0,0 +1,174 @@
+
+INSTANTIATE_TEST_SUITE_P(ShortSingleToken,
+    SplitTokenExtractorTest,
+    ::testing::ValuesIn(std::initializer_list<SplitTokenExtractorTestCase>{
```
Nit: As I remember, you don't need initializer_list, use std::span if it available in clickhouse std

Enmk replied:
it is C++20, not sure if all the features have landed to all compilers/platforms CH supports.

danlark1 replied:
You can test in CI, C++20 is on by default in ClickHouse, I checked, this should be supported already

Enmk replied:
well, I honestly fail to see how that is going to simplify the initialization of the test cases.

danlark1 commented on dbms/src/Storages/tests/gtest_SplitTokenExtractor.cpp:
```diff
@@ -0,0 +1,174 @@
+}
+
+// Helper to allow strings with embedded '\0' chars.
+#define BINARY_STRING(str) std::string{str, sizeof(str) - 1}
```
You can use c++17 suffix literal feature

https://en.cppreference.com/w/cpp/string/basic_string/operator%22%22s

https://gcc.godbolt.org/z/HY3nwV

Enmk replied:
Cool feature, but again, this is a C++20...

danlark1 replied:
Nope, this is a C++14 feature with C++20 improvements. Also in Clickhouse there is alread C++20

Enmk replied:
Okay, fixed

danlark1 commented on dbms/src/Storages/tests/gtest_SplitTokenExtractor.cpp:
```diff
@@ -0,0 +1,174 @@
+            {"abcdefghijklmnopqrstuvwxyz"}
+        },
+    })
+);
```
Also, please add some padding tests, it can easily be done with PaddedPOD array with garbage padding set and then resized

Enmk replied:
There is "implicit garbage" padding, but good point anyway, let's make sure that malicious padding does not break anything.

Enmk replied:
done

danlark1 reviewed (commented):
Sorry for many comments but I hope they help :)

Enmk committed:
Fixed builds, implementation and tests

* Builds shouldn't fail on platforms that do not support SSE2 and SSE4.2 and do not have corresponding headers.
* Updated tests to include malicious padding
* Fixed reporting tokens that cross or outside of data boundaries.

Enmk committed:
More tests and better token checks.

Enmk commented on dbms/Storages/MergeTree/MergeTreeIndexFullText.cpp:
```diff
@@ -606,8 +610,67 @@ bool SplitTokenExtractor::next(const char * data, size_t len, size_t * pos, size
+
+#if defined(__SSE4_2__)
+        // With the help of https://www.strchr.com/strcmp_and_strlen_using_sse_4.2
+        static const auto alnum_chars_ranges = _mm_set_epi8(0, 0, 0, 0, 0, 0, 0, 0,
```
great insight! fixed

Enmk commented on dbms/Storages/MergeTree/MergeTreeIndexFullText.cpp:
```diff
@@ -606,8 +610,67 @@ bool SplitTokenExtractor::next(const char * data, size_t len, size_t * pos, size
     while (*pos < len)
     {
+#if defined(__SSE2__)
+        // NOTE: we assume that `data` string is padded from the right with 15 bytes.
```
indeed, there was a bug, fixed and updated tests to detect it

danlark1 replied:
Can you point me where the bug was fixed?

Enmk replied:
https://github.com/ClickHouse/ClickHouse/pull/9968/files#diff-2fbd065f4a2e299d3c27719e489cc6d6R696-R699

Enmk committed:
Merge commit '73b0f8db8c327a1d63cc7ebcc56087a3f9866dae' into SIMD_tokenization

Enmk committed:
Minor test refactoring

* using string literal for binary strings
* sorted includes

Enmk committed:
Merge remote-tracking branch 'upstream/master' into SIMD_tokenization

Enmk commented:
Benchmark code and results:
https://gist.github.com/Enmk/2c72399733d52f1e20ea3df2051db2d1

Conclusion: 
* SSE2 && SSE4.2 versions are slower (~25%) if tokens are tiny (1 character)
* SSE2 && SSE4.2 are faster by ~10% on 2-char tokens
* larger the token, better the performance of SSE2 && SSE4.2 variants, up to ~44% (almost twice as fast) for tokens from 16 bytes and up.
* SSE4.2 variant is ALWAYS ~1%-5% faster than SSE2 variant.

Enmk commented on dbms/Storages/MergeTree/MergeTreeIndexFullText.cpp:
```diff
@@ -606,8 +610,67 @@ bool SplitTokenExtractor::next(const char * data, size_t len, size_t * pos, size
+        static const auto alnum_chars_ranges = _mm_set_epi8(0, 0, 0, 0, 0, 0, 0, 0,
+                '\xFF', '\x80', 'z', 'a', 'Z', 'A', '9', '0');
+        // Every bit represents if `haystack` character is in the ranges (1) or not(0)
+        const int result_bitmask = _mm_cvtsi128_si32(_mm_cmpestrm(alnum_chars_ranges, 8, haystack, haystack_length, _SIDD_CMP_RANGES));
```
https://github.com/ClickHouse/ClickHouse/pull/9968#issuecomment-611354294

danlark1 replied:
Thanks!

danlark1 commented:
Overall looks good to me, it would be also nice to test SSE2 vs SSE4.2 on AMD processors (I have a feeling that SSE2 might be faster there) but this is just an idea for future if somebody comes. @alexey-milovidov, your turn to review or merge

Enmk commented:
> Overall looks good to me, it would be also nice to test SSE2 vs SSE4.2 on AMD processors (I have a feeling that SSE2 might be faster there) but this is just an idea for future if somebody comes. @alexey-milovidov, your turn to review or merge

Sure, I'd be happy to see more test data. I hope that it is relatively simple to do that with the benchmark code I provided.

And thank you for the review!

alexey-milovidov reviewed (approved):
Ok.

alexey-milovidov commented on src/Storages/MergeTree/MergeTreeIndexFullText.cpp:
```diff
@@ -606,8 +615,67 @@ bool SplitTokenExtractor::next(const char * data, size_t len, size_t * pos, size
     while (*pos < len)
     {
+#if defined(__SSE2__)
+        // NOTE: we assume that `data` string is padded from the right with 15 bytes.
```
This is a bug - you cannot make such assumption, see https://github.com/ClickHouse/ClickHouse/issues/19233
